Source/JavaScriptCore/b3/air/AirCode.cpp

```diff
@@ -45,8 +45,8 @@ const char* const tierName = "Air ";
 static void defaultPrologueGenerator(CCallHelpers& jit, Code& code)
 {
     jit.emitFunctionPrologue();
+    AllowMacroScratchRegisterUsageIf allowScratch(jit, isARM64());
```

Comment:
On ARM64, we allow scratch register usage for this assertion to pass in `getCachedMemoryTempRegisterIDAndInvalidate` https://github.com/WebKit/WebKit/blob/aabfacb8cf9ca4780d20ce2cca233f7988e994e2/Source/JavaScriptCore/assembler/MacroAssemblerARM64.h#L4799 (triggered indirectly by `emitSave`).
I suppose this is because the callee save offsets might not satisfy the conditions in tryStoreWithOffset(https://github.com/WebKit/WebKit/blob/aabfacb8cf9ca4780d20ce2cca233f7988e994e2/Source/JavaScriptCore/assembler/MacroAssemblerARM64.h#L1899). This can happen if we shuffled the callee's stack args, and they end up as stack slots. The callee saves are then placed on top, and their offsets are too large for the conditions to pass.

Reply:
Hmm, this seems wrong, but then the existing code also seems wrong. I don't know why this was needed. It seems like the caller should be allowing macro scratch register usage. I am guessing that your assertion failed because AirGenerate doesn't have an AllowMacroScratchRegisterUsage scope? I wonder why it doesn't just create the scope. @Constellation might have some ideas.

Reply:
`Air::generateWithAlreadyAllocatedRegisters` has a `DisallowMacroScratchRegisterUsage` scope, I guess that explains the initial `allowScratch` here.

Reply:
Yeah, this seems super wrong and confusing